compiler/src/iree/compiler/Utils/EmbeddedDataDirectory.h

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_UTILS_EMBEDDEDDATADIRECTORY_H_
+#define IREE_COMPILER_UTILS_EMBEDDEDDATADIRECTORY_H_
+
+#include <mutex>
+#include "llvm/ADT/StringMap.h"
+
+namespace mlir::iree_compiler {
+
+// A string-to-StringRef map that acts as a virtual filesystem: the keys are
+// "filenames" and the values are file contents.
+class EmbeddedDataDirectory {
+public:
+  // Calls the given `callback` on a global singleton object, guarded by a
+  // global mutex.
+  //
+  // Only use this for use cases that require a global object, such as when
+  // exporting data between parts of the compiler that can't directly link to
+  // each other (e.g. from a plugin to outside of the plugin).
+  static void
+  withGlobal(llvm::function_ref<void(EmbeddedDataDirectory &)> callback) {
+    static EmbeddedDataDirectory dir;
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> lock(mutex);
```

Comment:
For educational purposes, when/what cases do we actually need it to be protected by global mutex? would we be adding ukernels (from plugin) while it being read (in GPULowerToUkernel) concurrently/at the same time?

Reply:
The current usage pattern does not actually need the mutex, and initially I omitted the mutex, while adding comments explaining why that was safe to do. At some point while refining these comments, I thought that I should invest less in fancy comments and more in hardening this interface against future mistakes: while my initial use case was safe without a mutex, I could envision future use cases being subtly different.

The initial usage pattern here is:
1. The only time where we ever mutate the directory (add files) is at ROCM plugin registration. This happens very early, on the main thread. If any other plugin starts adding more file, that will still be on the main thread, so there is no concern about concurrency here.
2. The only time where we ever read the directory (get files) is in a codegen pass (GPULowerToUKernels). This happens on LLVM worker threads, so this is definitely happening concurrently on multiple threads, but that's OK because these are purely read accesses. The `StringMap` method is `const`-qualified, so it would be a bit crazy if it actually mutated something, although as a `StringMap` detail that's not impossible (there is no comment that says that it wouldn't).

Having this `withGlobal` mechanism with the implicit mutex is a safeguard ensuring that people spreading this to more places with different usage patterns, don't need to worry about when exactly a mutex becomes necessary.

Part of my thinking was that since this helper is presenting itself as a "virtual filesystem" (`File` being part of its name), users will legitimately expect it to offer filesystem-like behavior w.r.t. concurrent accesses.

At the same time, the class itself is a very thin wrapper around the underlying `StringMap` and I liked to keep it that way.

So I liked Jakub's suggestion of `withGlobal`.

Reply:
Understood, thanks!